cloud_functions/gcs_event_based_ingest/gcs_ocn_bq_ingest/main.py

```diff
@@ -0,0 +1,378 @@
+# Copyright 2018 Google LLC.
+# This software is provided as-is, without warranty or representation
+# for any use or purpose.
+# Your use of it is subject to your agreement with Google.
+
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Background Cloud Function for loading data to BigQuery.
+"""
+import json
+from collections import deque
+from os import getenv
+from pathlib import Path
+from typing import Any
+from typing import Dict, List, Optional, Tuple
+from urllib.parse import urlparse
+import logging
+from time import monotonic, sleep
+
+from google.api_core.client_info import ClientInfo
+from google.cloud import bigquery, storage
+from google.cloud.exceptions import NotFound
+
+# 15TB per BQ load job.
+MAX_BATCH_BYTES = 15 * 10 ** 12
+MAX_URIS_PER_LOAD = 10 ** 4
+
+BASE_LOAD_JOB_CONFIG = {
+    "sourceFormat": "CSV",
+    "fieldDelimiter": "|",
+    "writeDisposition": "WRITE_APPEND",
+}
+
+DEFAULT_BQ_LOAD_STATE_TABLE = "bigquery_loads.serverless_bq_loads"
+
+DEFAULT_EXTERNAL_TABLE_DEFINITION = {
+    "sourceFormat": "PARQUET"
+}
+
+DEFAULT_JOB_LABELS = {
+    "component": "event-based-gcs-ingest",
+    "cloud-function-name": getenv("FUNCTION_NAME"),
+}
+
+BASE_LOAD_JOB_CONFIG = {
+    "sourceFormat": "CSV",
+    "fieldDelimiter": "|",
+    "writeDisposition": "WRITE_APPEND",
+    "labels": DEFAULT_JOB_LABELS,
+}
+
+# Will wait up to this polling for errors before exiting
+# This is to check if job fail quickly, not to assert the succeed
+# This may not be honored if longer than cloud function timeout
+# https://cloud.google.com/functions/docs/concepts/exec#timeout
+WAIT_FOR_JOB_SECONDS = 5
+
+
+def main(event: Dict, context):
+    # Set by Cloud Function Execution Environment
+    # https://cloud.google.com/functions/docs/env-var
+    project = getenv("GCP_PROJECT")
+
+
+    bucket_id, object_id = parse_notification(event)
+
+    success_filename = getenv("SUCCESS_FILENAME", "_SUCCESS")
+
+    # Exit eagerly if not a success file.
+    # TODO we can improve this with pub/sub message filtering once it supports
+    # a hasSuffix filter function (we can filter on hasSuffix successfile name)
+    #  https://cloud.google.com/pubsub/docs/filtering
+    if not object_id.endswith(f"/{success_filename}"):
```

Comment:
If a directory which has been previously loaded receives new data and a new `_SUCCESS` file, it seems like entire table will be reloaded. Is it safe to say this does not handle incremental loads? If so, perhaps the default disposition should be `WRITE_TRUNCTATE`

Reply:
I definitely want to handle incrementals.

I will add a note to the docs and a test for this as this will be a common use case. 
Incrementals should always happen in a prefix under the table prefix (e.g. a partition prefix) so that their behavior can be properly overwritten in a load.json .
e.g.
```
gs://bucket/dataset/table/_config/load.json          # load.json dictates WRITE_APPEND for historical
gs://bucket/dataset/table/parts/_config/load.json # load.json WRITE_TRUNCATE for all prefixes under parts/
```

However, if willing to pay for query slots for incrementals IMHO the most general way to solve incrementals should happen via external query as this provides richer DML semantics (e.g. `MERGE` into partition) than load job write disposition and access to the `_FILE_NAME` which can be used to. 

In the future as GCS > BQ DTS ongoing "mirroring" is supported, this asset would mostly be useful for orchestrating external query ingest.

Reply:
@ryanmcdowell can you see if the docs and regex system I've added in latest commits addresses your incremental loads use case? (of course it will be better once I've added integration test for this pattern)